src/Text/Trifecta/Parser.hs

```diff
@@ -129,14 +130,23 @@ instance Semigroup a => Semigroup (Parser a) where
   (<>) = liftA2 (<>)
   {-# INLINE (<>) #-}
 
+{- In order to silence '-Wnoncanonical-monoid-instances' we'd need to
+   change the type-signature to emulate the not-yet existing superclass
+   relationship:
+
+instance (Semigroup a, Monoid a) => Monoid (Parser a) where
+  mappend = (<>)
+  mempty = pure mempty
+-}
 instance Monoid a => Monoid (Parser a) where
```

Comment:
Again, I'm not concerned about making a breaking change here, since I think this change is warranted. I'd rather get it right and do a major version bump instead of having a hodgepodge of different constraints on different GHCs (aside from the one in https://github.com/ekmett/trifecta/pull/71#discussion_r118580014, which is unavoidable).